internal/crypto/src/raw_signature/validator.rs

```diff
@@ -0,0 +1,143 @@
+        return Some(validator);
+    }
+
+    None
+}
+
+/// Return a built-in signature validator for the requested signature
+/// algorithm as identified by OID.
+///
+/// Which validators are available may vary depending on the platform and
+/// which crate features were enabled.
+///
+/// TEMPORARILY PUBLIC: This will become `pub(crate)` once time stamp code moves
+/// into c2pa-crypto
+pub fn validator_for_sig_and_hash_algs(
+    sig_alg: &Oid,
+    hash_alg: &Oid,
+) -> Option<Box<dyn RawSignatureValidator>> {
+    if sig_alg.as_ref() == RSA_OID.as_bytes()
+        || sig_alg.as_ref() == SHA256_WITH_RSAENCRYPTION_OID.as_bytes()
+        || sig_alg.as_ref() == SHA384_WITH_RSAENCRYPTION_OID.as_bytes()
+        || sig_alg.as_ref() == SHA512_WITH_RSAENCRYPTION_OID.as_bytes()
+    {
+        // TO REVIEW: Do we need any of the RSA-PSS algorithms for this use case?
+
+        #[cfg(feature = "openssl")]
+        if let Some(validator) =
+            crate::openssl::validators::validator_for_sig_and_hash_algs(sig_alg, hash_alg)
+        {
+            return Some(validator);
+        }
+
+        // Not sure yet if we'll need legacy validators for WASM.
+        // #[cfg(all(target_arch = "wasm32", not(target_os = "wasi")))]
+        // if let Some(validator) =
+        //     crate::webcrypto::validators::validator_for_sig_and_hash_algs(sig_alg, hash_alg)
+        // {
+        //     return Some(validator);
+        // }
+    } else if sig_alg.as_ref() == EC_PUBLICKEY_OID.as_bytes()
+        || sig_alg.as_ref() == ECDSA_WITH_SHA256_OID.as_bytes()
+        || sig_alg.as_ref() == ECDSA_WITH_SHA384_OID.as_bytes()
+        || sig_alg.as_ref() == ECDSA_WITH_SHA512_OID.as_bytes()
+    {
+        if hash_alg.as_ref() == SHA256_OID.as_bytes() {
+            return validator_for_signing_alg(SigningAlg::Es256);
+        } else if hash_alg.as_ref() == SHA384_OID.as_bytes() {
+            return validator_for_signing_alg(SigningAlg::Es384);
+        } else if hash_alg.as_ref() == SHA512_OID.as_bytes() {
+            return validator_for_signing_alg(SigningAlg::Es512);
+        }
+    } else if sig_alg.as_ref() == ED25519_OID.as_bytes() {
+        return validator_for_signing_alg(SigningAlg::Ed25519);
+    }
+
+    None
+}
+
+/// Describes errors that can be identified when validating a raw signature.
+#[derive(Debug, Eq, Error, PartialEq)]
+#[non_exhaustive]
+pub enum RawSignatureValidationError {
+    /// The signature does not match the provided data or public key.
+    #[error("the signature does not match the provided data or public key")]
+    SignatureMismatch,
+
+    /// An error was reported by the OpenSSL native code.
+    ///
+    /// NOTE: We do not directly capture the OpenSSL error itself because it
+    /// lacks an Eq implementation. Instead we capture the error description.
+    #[cfg(feature = "openssl")]
+    #[error("an error was reported by OpenSSL native code: {0}")]
+    OpenSslError(String),
```

Comment:
I think this is a very good choice.

Reply:
Should we allow the Error to return LogItem to get the detailed error info?

Reply:
@mauricefisher64 I saw this error case a few times while debugging the new code and I was pretty pleased with the description that was provided. It included the OpenSSL file name and line number.